File: the.py
def glc(a):
    from datetime import datetime
    import math
    start = datetime.now()
    # -----------------------
    def is_prime(g):
        mod = 2
        output = True
        try:
            if (g >= 2 and g % 1 == 0):
                limit = int(math.sqrt(g))
                while (mod <= limit):
                    if (g % mod == 0):
                        output = False
                        break
                    mod += 1
            else:
                output = False
        except:
            output = False
        return output
    # -----------------------
    def test_input(b):
        output = False
        i = 0
        not_positive_int = False
        # -----------------------
        def invalid_input(a):
            return (a < 0 or a % 1 != 0 or a > 1e6)
        # -----------------------
        if isinstance(b, list):
            if len(b) > 1:
                if len(b) <= 20:
                    while (i < len(b)):
                        if (invalid_input(b[i])):
                            not_positive_int = True
                            break
                        i += 1
                    if not not_positive_int:
                        output = True
        return output
    # -----------------------
    def gl_calc(a):
        arr = a[:]
        arr.sort()
        arr = list(reversed(arr))
        gcf = []
        lcm = []
        divisor = 2
        buffer = arr[:]
        # -----------------------
        def check_divisor(flag):
            can_be_divided = 0
            output = False
            for i, v in enumerate(buffer):
                if (v / float(divisor)) % 1 == 0:
                    can_be_divided += 1
                    output = True
                    if (flag is True):
                        buffer[i] = int(v / float(divisor))
            return [output, can_be_divided]
        # -----------------------
        def division_complete():
            c = 0;
            for v in buffer:
                c += v
            return (c == len(buffer))
        # -----------------------
        while True: # Outer loop
            if not check_divisor(False)[0]:
                while True: # Inner loop
                    divisor += 1
                    if is_prime(divisor): # Inner loop limit
                        break
            chk_div = check_divisor(True)
            if (chk_div[0] is True):
                if (chk_div[1] > 0):
                    lcm.append(divisor)
                if (chk_div[1] == len(buffer)):
                    gcf.append(divisor)
            if division_complete(): # Outer loop limit
                break
            # -----------------------
        def add_one(a):
            if len(a) == 0:
                a.append(1)
            else:
                a.insert(0, 1)
            return a
        # -----------------------
        def mul(a):
            mult = 1
            for v in a:
                if v is not None:
                    mult *= v
            return mult
        # -----------------------
        return {
            "input": a,
            "output": {
                "gcf_array": add_one(gcf),
                "lcm_array": add_one(lcm),
                "gcf": mul(gcf),
                "lcm": mul(lcm)
            }
        }
        # -----------------------
    gl_result = "Error: invalid input"
    if (test_input(a)):
        gl_result = gl_calc(a)
        if isinstance(gl_result, dict):
            end = datetime.now()
            tt = end - start
            tt = int(tt.seconds) + int(tt.microseconds / 1000)
            gl_result["duration"] = str(tt) + " ms"
            # -----------------------
    return gl_result

File: the.js
function glc(a) {
    "use strict";
    var start = Date.now();
    var end;
    var test_input = function (b) { // test input array validity
        var output = false;
        var i = 0;
        var not_positive_int = false;
        var invalid_input = function (a) {
            return (a <= 0 || a % 1 !== 0 || a > 1e6);
        };
        if (b !== undefined) {
            if (b.constructor === Array) {
                if (b.length > 1) {
                    if (b.length <= 20) { // maximum array length is 20
                        while (i < b.length) {
                            if (invalid_input(b[i])) {
                                not_positive_int = true;
                                break;
                            }
                            i += 1;
                        }
                        if (!not_positive_int) {
                            output = true;
                        }
                    }
                }
            }
        }
        return output;
    };

    var gl_calc = function (a) { // GCF & LCM CALCULATION
        var is_prime = function (g) { // prime checker
            var limit;
            var mod = 2;
            var output = true;
            if (g >= 2 && g % 1 === 0) {
                limit = Math.floor(Math.sqrt(g));
                while (mod <= limit) {
                    if (g % mod === 0) {
                        output = false;
                        break;
                    }
                    mod += 1;
                }
            } else {
                output = false;
            }
            return output;
        };

        var arr = a.slice().sort().reverse(); // input array is sorted downward (high → low)
        var max = Math.floor(Math.sqrt(arr[0])); // limit for loop
        var gcf = [];
        var lcm = [];
        var divisor = 2;
        var buffer = arr.slice(); // division result
        var chk_div;

        var check_divisor = function (flag) {
            var can_be_divided = 0;
            var output = false;
            buffer.forEach(function (v, i) { // find out whether each element can be divided or not
                if ((v / divisor) % 1 === 0) {
                    if (flag) {
                        buffer[i] = v / divisor; // update buffer array
                    }
                    can_be_divided += 1;
                    output = true;
                }
            });
            return [output, can_be_divided];
        };

        var gen_divisor = function () {
            if (!check_divisor(false)[0]) { // false → check current divisor WITHOUT updating buffer array
                do {
                    divisor += 1; // increment divisor
                } while (!is_prime(divisor) && divisor <= max);
            }
        };

        var gen_check_divisor = function () {
            gen_divisor(); // generate a prime number as current divisor
            chk_div = check_divisor(true); // true → check current divisor AND update buffer array
            if (chk_div[0] === true) { // true as chk_div[0] → at least one element can be divided
                if (chk_div[1] > 0) {
                    lcm.push(divisor);
                }
                if (chk_div[1] === buffer.length) { // chk_div[1] → can_be_divided (ALL can be divided)
                    gcf.push(divisor);
                }
            }
        };

        var division_complete = function () { // check if array has only integer 1 for each element
            var c = 0;
            buffer.forEach(function (v) {
                c += v;
            });
            return (c === buffer.length);
        };

        // ----------------------------------------- //
        do { // MAIN LOOP                            //
            gen_check_divisor();                     //
        } while (!division_complete(buffer));        //
        // ----------------------------------------- //

        var add_one = function (a) { // 1
            if (!a.length) {
                a.push(1);
            } else {
                a.unshift(1);
            }
            return a;
        };

        var mul = function (a) { // multiplications of array elements
            var mult = 1;
            a.forEach(function (v) {
                mult *= v;
            });
            return mult;
        };

        return { // FINAL RESULT
            input: a,
            output: {
                gcf_array: add_one(gcf),
                lcm_array: add_one(lcm),
                gcf: mul(gcf),
                lcm: mul(lcm)
            }
        };
    };

    var gl_result = "Error: invalid input → must be array of positive integers → max each 1,000,000 (one million)";
    if (test_input(a)) {
        gl_result = gl_calc(a); // RUN CALCULATION
    }
    if (gl_result.constructor === Object) {
        end = Date.now();
        gl_result.duration = (end - start) + " ms";
    }
    return gl_result;
}
